pkg/controller/atlasproject/encryption_at_rest.go

```diff
@@ -34,6 +45,116 @@ func ensureEncryptionAtRest(ctx *workflow.Context, projectID string, project *md
 	return workflow.OK()
 }
 
+func readEncryptionAtRestSecrets(kubeClient client.Client, service *workflow.Context, encRest *mdbv1.EncryptionAtRest, parentNs string) error {
+	if encRest == nil {
+		return nil
+	}
+
+	if encRest.AwsKms.Enabled != nil && *encRest.AwsKms.Enabled && encRest.AwsKms.SecretRef.Name != "" {
+		watchObj, err := readAndFillAWSSecret(kubeClient, parentNs, &encRest.AwsKms)
+		service.AddResourcesToWatch(*watchObj)
+		if err != nil {
+			return err
+		}
+	}
+
+	if encRest.GoogleCloudKms.Enabled != nil && *encRest.GoogleCloudKms.Enabled && encRest.GoogleCloudKms.SecretRef.Name != "" {
+		watchObj, err := readAndFillGoogleSecret(kubeClient, parentNs, &encRest.GoogleCloudKms)
+		service.AddResourcesToWatch(*watchObj)
+		if err != nil {
+			return err
+		}
+	}
+
+	if encRest.AzureKeyVault.Enabled != nil && *encRest.AzureKeyVault.Enabled && encRest.AzureKeyVault.SecretRef.Name != "" {
+		watchObj, err := readAndFillAzureSecret(kubeClient, parentNs, &encRest.AzureKeyVault)
+		service.AddResourcesToWatch(*watchObj)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func readAndFillAWSSecret(kubeClient client.Client, parentNs string, awsKms *mdbv1.AwsKms) (*watch.WatchedObject, error) {
+	fieldData, watchObj, err := readSecretData(kubeClient, awsKms.SecretRef, parentNs, "CustomerMasterKeyID", "Region", "RoleID")
+	if err != nil {
+		return watchObj, err
+	}
+
+	fillStructFields(fieldData, awsKms)
+	return watchObj, nil
+}
+
+func readAndFillGoogleSecret(kubeClient client.Client, parentNs string, gkms *mdbv1.GoogleCloudKms) (*watch.WatchedObject, error) {
+	fieldData, watchObj, err := readSecretData(kubeClient, gkms.SecretRef, parentNs, "ServiceAccountKey", "KeyVersionResourceID")
+	if err != nil {
+		return watchObj, err
+	}
+	fillStructFields(fieldData, gkms)
+	return watchObj, err
+}
+
+func readAndFillAzureSecret(kubeClient client.Client, parentNs string, azureVault *mdbv1.AzureKeyVault) (*watch.WatchedObject, error) {
+	fieldData, watchObj, err := readSecretData(kubeClient, azureVault.SecretRef, parentNs, "ClientID", "AzureEnvironment", "SubscriptionID", "ResourceGroupName", "KeyVaultName", "KeyIdentifier")
+	if err != nil {
+		return watchObj, err
+	}
+	fillStructFields(fieldData, azureVault)
+	return watchObj, err
+}
+
+// Fills public fields for the "ptrStruct" using field names from "data" map K -> V string/string. "ptrStruct" must be a pointer to a struct
+func fillStructFields(data map[string]string, ptrStruct any) {
+	ptr := reflect.ValueOf(ptrStruct).Elem()
+	if ptr.Kind() != reflect.Struct {
+		return
+	}
+
+	for k, v := range data {
+		field := ptr.FieldByName(k)
+		if field.IsValid() && field.CanSet() && field.Type() == reflect.TypeOf(v) {
+			field.Set(reflect.ValueOf(v))
+		}
+	}
+}
+
+// Return all requested field from a secret
+func readSecretData(kubeClient client.Client, res common.ResourceRefNamespaced, parentNamespace string, fieldNames ...string) (map[string]string, *watch.WatchedObject, error) {
```

Comment:
ok, I see that this `reasSecretData` does some useful validation, while reading the secret. But we could do the validation alone and still have plain explicit conversions.